### ISSUE TITLE
Coveralls integration stopped working around 2018-09-20

### ISSUE THREAD

**Issue by maxbennedich** [JuliaCI/Coverage.jl#175]
This seems to be a global issue, I've checked several packages, and they all started failing around Sept 20. This is what happens during submission of coverage:
```
Submitting data to Coveralls...
┌ Warning: Error in @async writebody task
│   exception =
│    MethodError: no method matching write(::HTTP.Streams.Stream{HTTP.Messages.Response,HTTP.ConnectionPool.Transaction{MbedTLS.SSLContext}}, ::Pair{String,HTTP.Multipart{Base.GenericIOBuffer{Array{UInt8,1}}}})
...
 [21] #submit#1(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::Array{FileCoverage,1}) at /home/travis/.julia/packages/Coverage/25tif/src/coveralls.jl:91
 [22] submit(::Array{FileCoverage,1}) at /home/travis/.julia/packages/Coverage/25tif/src/coveralls.jl:60
```
See example of full stack trace in [this build](https://travis-ci.org/JuliaData/DataFrames.jl/jobs/433126893#L1380).

Codecov still works.

**Comment by vchuravy** [comment]
x-ref: https://github.com/JuliaWeb/HTTP.jl/issues/245#issuecomment-414847061

**Comment by samoconnor** [comment]
Passing a `HTTP.Form` as the body and setting the `Content-Type` should do the trick:
```julia
using HTTP
using JSON

makebody(data::Dict) =
    HTTP.Form(
    Dict("json_file" => HTTP.Multipart("json_file", IOBuffer(JSON.json(data)),
                                       "application/json")))

url = "https://coveralls.io/api/v1/jobs"

data = Dict("service_job_id"    => "APPVEYOR_JOB_ID",
            "service_name"      => "appveyor",
            "source_files"      => "xyz",
            "repo_token"        => "REPO_TOKEN")

body = makebody(data)
headers = ["Content-Type" => "multipart/form-data; boundary=$(body.boundary)"]

req = HTTP.post(url, headers, body, verbose=3)
```
```
DEBUG: 2018-10-02T08:04:05.335 6516 ➡️  "POST /api/v1/jobs HTTP/1.1\r\n" (write)
DEBUG: 2018-10-02T08:04:05.361 6516 ➡️  "Content-Type: multipart/form-data; boundary=7ca059ffd29b644fb400a866be6b2ed8\r\n" (write)
DEBUG: 2018-10-02T08:04:05.361 6516 ➡️  "Host: coveralls.io\r\n" (write)
DEBUG: 2018-10-02T08:04:05.362 6516 ➡️  "Content-Length: 291\r\n" (write)
DEBUG: 2018-10-02T08:04:05.362 6516 ➡️  "\r\n" (write)
DEBUG: 2018-10-02T08:04:05.666 ecee ➡️  "--7ca059ffd29b644fb400a866be6b2ed8\r\n" (unsafe_write)
DEBUG:                                 "Content-Disposition: form-data; name=\"json_file\"; filename=\"json_file\"\r\n"
DEBUG:                                 "Content-Type: application/json\r\n"
DEBUG:                                 "\r\n"
DEBUG:                                 "{\"repo_token\":\"REPO_TOKEN\",\"service_name\":\"appveyor\",\"service_job_id\":\"APPVEYOR_JOB_ID\",\"source_files\":\"xyz\"}\r\n"
DEBUG:                                 "--7ca059ffd29b644fb400a866be6b2ed8--\r\n"
```

**Comment by samoconnor** [comment]
Note: this works because `HTTP.Form` is a readable `IO`. So the generic `HTTP.request` support for sending a body from an `IO` is used.

**Comment by samoconnor** [comment]
I think the best thing to do is add something like this to HTTP.jl:
```julia
content_type(f::Form) = "Content-Type" =>
                        "multipart/form-data; boundary=$(f.boundary)"

post(url, f::Form; kw...) = post(url, Header[], f; kw...)

function post(url, headers, f::Form; kw...)
    setheader(headers, content_type(f))
    request("POST", url, headers, f; kw...)
end
```
Then client code will be able to do:
```julia
req = HTTP.post(url, HTTP.Form(data))
```

**Comment by samoconnor** [comment]
See https://github.com/JuliaWeb/HTTP.jl/commit/01a2cff816e3cd6095990b80167f216cb26e906c

**Comment by ronisbr** [comment]
Hi guys! Is there any workaround for this bug ?

**Comment by fingolfin** [comment]
This is fixed by PR #179

**Comment by cdsousa** [comment]
This seems not be corrected yet, at least for DataFrames.jl, see https://travis-ci.org/JuliaData/DataFrames.jl/jobs/443235134 line 1000.

**Comment by ararslan** [comment]
Once the next version of Coverage is merged into METADATA we should be good.

**Comment by ararslan** [comment]
Seems to have been un-fixed with HTTP 0.8: https://travis-ci.org/JuliaDiffEq/ModelingToolkit.jl/jobs/483694231#L681-L691
